aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -1352,6 +1315,84 @@ TORCH_IMPL_FUNC(_scatter_reduce_structured_cpu)
   });
 }
 
+void index_fill_out_helper(
+    const Tensor& self,
+    int64_t dim,
+    const Tensor& index,
+    const Scalar& value,
+    const Tensor& out) {
+
+  if (index.numel() == 0) return;
+  if (!out.is_same(self)){
+    out.copy_(self);
+  }
+
+  // Handle the case when `self` is 0-dim
+  Tensor out_nonzero_dim = (out.dim() == 0) ? out.unsqueeze(-1) : out;
+
+  // Prepare `index` for TensorIterator.
+  // It is restrided to be broadcastable over `self` in TensorIterator.
+  auto index_sizes = std::vector<int64_t>(out_nonzero_dim.dim(), 1);
+  auto index_strides = std::vector<int64_t>(out_nonzero_dim.dim(), 0);
+  index_sizes[dim] = index.numel();
+  index_strides[dim] = (index.dim() > 0) ? index.stride(0) : 1; // `index` is 1d or scalar
+  auto index_restrided = index.as_strided(index_sizes, index_strides);
+
+  // Prepare `out` for TensorIterator.
+  // Restride `out` to not advance in dimension `dim`.
+  // We do not use squash_dim here because `index` will
+  // need to advance in this dimension.
+  // Note that out_sizes[dim] is set to index.numel().
+  // This is done so that out_sizes[dim] and index_sizes[dim]
+  // match as required by TensorIterator (input shape should
+  // strictly broadcast over output shape, i.e.
+  // output.shape[i] >= input.shape[i] for i in range(dims)).
+  auto out_sizes = out_nonzero_dim.sizes().vec();
+  auto out_strides = out_nonzero_dim.strides().vec();
+  out_sizes[dim] = index.numel();
+  out_strides[dim] = 0;
+  auto out_restrided = out_nonzero_dim.as_strided(out_sizes, out_strides);
+
+  auto iter = TensorIteratorConfig()
+    // We do not check for overlap because `self` is restrided
+    // with zero stride. Zero strides trigger memory overlap assert
+    // within TensorIterator.
+    .set_check_mem_overlap(false)
+    .check_all_same_dtype(false)
+    .resize_outputs(false)
+    .add_output(out_restrided)
+    .add_input(index_restrided)
+    .build();
```

Comment:
Probably, we can inherit from TensorIteratorBase and build it inside the meta block and thus check dtypes/sizes/devices there (which is missing right now).